File: 1306-JumpGameIII/1306-JumpGameIII.py
# Last updated: 2/2/2026, 3:36:37 PM
1class Solution:
2    def canReach(self, arr: List[int], start: int) -> bool:
3        g = [[] for _ in range(len(arr))]
4        target = set()
5        for i, n in enumerate(arr):
6            if n == 0:
7                target.add(i)
8            if 0 <= i - n:
9                g[i].append(i - n)
10            if i + n < len(arr):
11                g[i].append(i + n)
12                
13        q = deque([start])
14        v = [-1] * len(arr)
15        while q:
16            curr = q.pop()
17            for i in g[curr]:
18                if i in target:
19                    return True
20                elif v[i] == -1:
21                    q.append(i)
22                    v[i] = 0
23
24        return False

File: 1306-JumpGameIII/1306-JumpGameIII.cpp
// Last updated: 5/18/2026, 4:07:11 PM
1class Solution {
2    bool helper(vector<int>& arr, int idx, vector<bool>& visited) {
3        if (idx < 0 || idx >= arr.size()) return false;
4        if (visited[idx]) return false;
5        if (arr[idx] == 0) return true;
6
7        visited[idx] = true;
8
9        bool right = helper(arr, idx + arr[idx], visited);
10        bool left = helper(arr, idx - arr[idx], visited);
11
12        if (right || left) return true;
13
14        return false;
15    }
16public:
17    bool canReach(vector<int>& arr, int start) {
18        vector<bool> visited(arr.size(), false);
19        return helper(arr, start, visited);
20    }
21};